Firestore/Example/Tests/Local/FSTLocalStoreTests.mm

```diff
@@ -723,12 +734,14 @@ - (void)testCanExecuteMixedCollectionQueries {
 
   [self.localStore locallyWriteMutations:@[ FSTTestSetMutation(@"foo/bonk", @{@"a" : @"b"}) ]];
 
-  FSTDocumentDictionary *docs = [self.localStore executeQuery:query];
-  XCTAssertEqualObjects([docs values], (@[
-                          FSTTestDoc("foo/bar", 20, @{@"a" : @"b"}, NO),
-                          FSTTestDoc("foo/baz", 10, @{@"a" : @"b"}, NO),
-                          FSTTestDoc("foo/bonk", 0, @{@"a" : @"b"}, YES)
-                        ]));
+  DocumentDictionary docs = [self.localStore executeQuery:query];
+  XCTAssertEqual(docs.size(), 3);
+  XCTAssertEqualObjects(
+      (@[ docs.begin()->second, (++docs.begin())->second, docs.rbegin()->second ]), (@[
```

Comment:
I'm not a fan of this style of indexing. Could we instead create a helper, maybe something like this:

```
vector<FSTTestDoc> getValues(DocumentDIctionary docDict) {
  vector result;
  for (FSTTestDoc doc : docDict) result.push_back(doc);
  return result;
}
```

Then this code would become something like:
```
XCTAssertEqualObjects(getValues(docs), (@[
  FSTTestDoc(...),
  ...
]));
```

(Maybe s/vector/NSArray or whatever to make it work.)

Reply:
+1.

Reply:
Collecting an iterable thing up into a collection for testing seems like a general testing utility we should just have.

There's a [version already submitted](https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/core/test/firebase/firestore/immutable/array_sorted_map_test.cc#L137) that works with collections that define a `value_type` member.

This could also be adapted to just iterable ranges with something dependent on `decltype` of the result of `begin()`, however, I can't find a more concise way to express that than this:

```c++
template <typename Range>
auto ToVector(const Range& range)
    -> std::vector<
        typename std::remove_const<
            typename decltype(range.begin())::value_type>::type> {
  std::vector<
        typename std::remove_const<
            typename decltype(range.begin())::value_type>::type>
            result{range.begin(), range.end()};
  return result;
}
```

... which is a readability disaster but could possibly be worth it with a meaty enough comment above it.

Secondarily, the `values` projection of the DocumentDictionary is available via the [`value_view` iterator adaptor](https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/core/src/firebase/firestore/util/iterator_adaptors.h#L674).

Using that we can define a helper like this:
```c++

template <typename Map>
typename std::vector<typename std::remove_const<typename Map::mapped_type>::type>
Values(const Map& map) {
  return ToVector(util::value_view(map));
}
```

This means that with the above you could do
```c++
std::vector<FSTDocument*> expected{
    FSTTestDoc("foo/bar", 20, @{@"a" : @"b"}, NO),
    FSTTestDoc("foo/baz", 10, @{@"a" : @"b"}, NO),
    FSTTestDoc("foo/bonk", 0, @{@"a" : @"b"}, YES)
};
ASSERT_EQ(expected, testutil::Values(docs));
```

Reply:
A couple ideas on simplification:
* repeating `vector` type can be omitted using uniform initialization `return {range.begin(), range.end());`;
* return type declaration can be split in two parts by introducing a helper alias.
What I got:
```cpp
namespace {
  template <typename T>
  using Vec = std::vector<typename std::remove_const<T>::type>;
}

template <typename Range>
auto ToVector(const Range& range) -> 
    Vec<typename decltype(range.begin())::value_type> {
  return {range.begin(), range.end()};    
}
```

Reply:
Another idea - introducing an overload taking two iterators, though it's probably not an improvement:
```cpp
namespace {
  template <typename T>
  using Vec = std::vector<typename std::remove_const<T>::type>;
}

template <typename Iter>
auto ToVector(Iter b, Iter e) -> Vec<typename Iter::value_type> {
  return {b, e};    
}

template <typename Range>
auto ToVector(const Range& range)
    -> decltype(ToVector(range.begin(), range.end())) {
  return ToVector(range.begin(), range.end());
}
```

Reply:
Using uniform initialization is a big win already.

Splitting the return type as you've done doesn't work that well though because it introduces Vec<T> into every translation unit that includes the header declaring `ToVector`.

Reply:
Ah, remove_const was required when I was trying to use `decltype(*range.begin())`. Removing it and using uniform initialization leads to this:

```c++
template <typename Range>
auto ToVector(const Range& range)
    -> std::vector<typename decltype(range.begin())::value_type> {
  return {range.begin(), range.end()};
}
```

Which I can almost read! :-)